### PR TITLE
Set platform specific cache key

### PR DESCRIPTION
**Pull request by hans2103**
closed #6997 in favor of this one. Started all over again.

> problem: Joomla! makes no difference in devices when caching a page.
> This pull request makes an end to it and adds a prefix to the cache ID. 
> This pull request uses the Joomla! default browser detection implementation.

To test this PR you have to create a template override with platform specific code in it. For example an override on com_content article. 
Example code:

```
        jimport('joomla.application.web.client');
        $webclient = new JApplicationWebClient();

        if ($webclient->mobile)
        {
            echo 'Joomla! rocks';
        }
```

On mobile you will see the text "Joomla! rocks"
The same page on desktop or tablet you will not see the text "Joomla! rocks".

**before this patch && caching enabled:**
- The first platform (mobile or desktop) will initiate the cache.
- When the other platform asks for this page it will get the cached page for the other platform. (desktop will get mobile and vice versa)

**after this patch && caching enabled:**
-  enable the parameter "Platform specific caching" in the Joomla! configuration
- Both platforms are cached on visit by that platform. 
- When the other platform asks for this page it will get the cached page for that platform. (desktop gets cached version of desktop)

With this PR you can create platform specific pages AND use cache. (dance)
Special thanks to @roland-d for helping me with this pr

### CODE REVIEW

**Comment by n9iels** [comment]
I have tested this item :white_check_mark: successfully on 8c1c750d8652137af6023d8f37093d5e73412b13<br /><br />After enabled the parameter "Platform specific caching" in the Joomla! configuration, the patch works fine for me <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/8055">issues.joomla.org/joomla-cms/8055</a>.</sub>

**Comment by hans2103** [comment]
@n9iels added "enable the parameter "Platform specific caching" in the Joomla! configuration" to the test instructions. Thx

File: libraries/joomla/cache/cache.php [hunk 1]
```diff
@@ -751,6 +751,33 @@ public static function makeId()
...
+	 * @since 3.5
+	 */
+	public static function getPlatformPrefix()
+	{
+		$conf = JFactory::getConfig();
+
+		// No prefix when Global Config is set to no platfom specific prefix
+		if (!$conf->get('cache_platformprefix', '0'))
+		{
+			return;
```
Comment (Fedik) [inline libraries/joomla/cache/cache.php]: you have declared ` * @return  string`, so here need `return '';`, 
same at the end of this method ...

**Comment by pe7er** [comment]
@test successful 

In Global Configuration I've enabled Conservative Cache. 
I've created a template override /templates/protostar/html/com_content/article/default.php with the following code:
```php
<?php 
jimport('joomla.application.web.client');
$webclient = new JApplicationWebClient();
if ($webclient->mobile)
{
	echo 'Joomla! mobile rocks';
}else{
	echo 'Joomla! desktop rocks';
}
?>
```
After applying the patch the same issue occurred: if I first visit the page with mobile phone, then the desktop will also see "Joomla! mobile rocks".  **After setting "Platform specific caching" to "Yes"** and emptying Joomla's cache, the desktop showed 'Joomla! desktop rocks' and the mobile displayed 'Joomla! mobile rocks'.

Not that **I had to use Google Chrome Inspector > Inspect Element** > click on the **mobile phone** & select a device **to properly display 'Joomla! mobile rocks'**. Just re-sizing the browser to mobile phone size did not work.

**Comment by pe7er** [comment]
I have tested this item :white_check_mark: successfully on 556c98a46dc4c7eb08601e670f93083615bb8ee2<br /><br />I've tested this PR successfully but because I used github, the system does not see it as tested.
Therefore I've resubmitted my successful test report http://issues.joomla.org/tracker/joomla-cms/8055#event-104374 via issues.joomla.org<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/8055">issues.joomla.org/joomla-cms/8055</a>.</sub>

**Comment by RemcoJanssen** [comment]
I have tested this item :white_check_mark: successfully on 556c98a46dc4c7eb08601e670f93083615bb8ee2<br /><br />Worked perfectly<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/8055">issues.joomla.org/joomla-cms/8055</a>.</sub>

**Comment by zero-24** [comment]
RTC for 3.5 thanks<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/8055">issues.joomla.org/joomla-cms/8055</a>.</sub>

**Comment by roland-d** [comment]
Thanks all. @rdeutz merged this into 3.5-dev with commit aea70b2

**Comment by fevangelou** [comment]
a) Has this been tested with the (system) page cache plugin?

b) If the browser detection is being done with PHP (which I assume it is, since you can the J API), you understand that this will have no effect when used with a reverse caching proxy like Varnish, right?

**Comment by roland-d** [comment]
@fevangelou A little late ;)

I don't remember if this has been tested with the page cache plugin.

If it has no effect than at least it works as it used to work before, feel free to post a pull request that deals with those cases as well. By default this option is turned off.

**Pull request closed without merging**